src/lib/tcp/src/seq.rs

```diff
@@ -0,0 +1,270 @@
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub(crate) struct Seq(u32);
+
+// We don't implement `From<u32>` or `Deref` since it makes it easier to accidentally mix up
+// operating on a `u32` instead of a `Seq`. For example this can cause bugs if accidentally adding
+// using `<u32 as Add<u32>>` instead of `<Seq as Add<u32>>`, which has a different wrapping
+// behaviour. We don't implement `PartialOrd` or `Ord` since there is no ordering relation between
+// arbitrary sequence numbers modulo 2^32.
+static_assertions::assert_not_impl_any!(Seq: PartialOrd, Ord, From<u32>, std::ops::Deref);
+
+impl Seq {
+    #[inline]
+    pub fn new(x: u32) -> Self {
+        Self(x)
+    }
+}
+
+impl std::fmt::Debug for Seq {
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl std::fmt::Display for Seq {
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl From<Seq> for u32 {
+    #[inline]
+    fn from(x: Seq) -> Self {
+        x.0
+    }
+}
+
+impl std::ops::Add<u32> for Seq {
+    type Output = Self;
+
+    fn add(self, offset: u32) -> Self::Output {
+        Self::new(self.0.wrapping_add(offset))
+    }
+}
+
+impl std::ops::Sub<u32> for Seq {
+    type Output = Self;
+
+    fn sub(self, offset: u32) -> Self::Output {
+        Self::new(self.0.wrapping_sub(offset))
+    }
+}
+
+impl std::ops::Sub for Seq {
+    type Output = u32;
+
+    fn sub(self, other: Self) -> Self::Output {
+        self.0.wrapping_sub(other.0)
+    }
+}
+
+impl std::ops::AddAssign<u32> for Seq {
+    fn add_assign(&mut self, offset: u32) {
+        self.0 = self.0.wrapping_add(offset);
+    }
+}
+
+impl std::ops::SubAssign<u32> for Seq {
+    fn sub_assign(&mut self, offset: u32) {
+        self.0 = self.0.wrapping_sub(offset);
+    }
+}
+
+/// A half-open range of sequence numbers modulo 2<sup>32</sup> bounded inclusively by `start` and
+/// exclusively by `end`. The starting position can be greater than the ending position.
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub(crate) struct SeqRange {
+    /// Inclusive starting position.
+    pub start: Seq,
+    /// Exclusive ending position.
+    pub end: Seq,
+}
+
+impl std::fmt::Debug for SeqRange {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.start.fmt(f)?;
+        write!(f, "..")?;
+        self.end.fmt(f)?;
+        Ok(())
+    }
+}
+
+impl SeqRange {
+    #[inline]
+    pub fn new(start: Seq, end: Seq) -> Self {
+        Self { start, end }
+    }
+
+    #[inline]
+    pub fn len(&self) -> u32 {
+        self.end - self.start
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns `true` if the sequence number is contained within this half-open range.
+    #[inline]
+    pub fn contains(&self, seq: Seq) -> bool {
+        seq - self.start < self.end - self.start
```

Comment:
Took me a little to see how this works. Optional, maybe clearer as:

```
seq - self.start < self.len()
```
or
```
SeqRange::new(self.start, seq).len() < self.len()
```

Reply:
Added the last version.